cores/esp32/esp32-hal-touch.c

```diff
@@ -12,219 +12,242 @@
 {
-#if CONFIG_IDF_TARGET_ESP32
-    uint32_t pad_intr = READ_PERI_REG(SENS_SAR_TOUCH_CTRL2_REG) & 0x3ff;
-    uint32_t rtc_intr = READ_PERI_REG(RTC_CNTL_INT_ST_REG);
-    uint8_t i = 0;
+#if SOC_TOUCH_VERSION_1         // ESP32 
+    uint32_t pad_intr = touch_pad_get_status();
     //clear interrupt
-    WRITE_PERI_REG(RTC_CNTL_INT_CLR_REG, rtc_intr);
-    SET_PERI_REG_MASK(SENS_SAR_TOUCH_CTRL2_REG, SENS_TOUCH_MEAS_EN_CLR);
-
-    if (rtc_intr & RTC_CNTL_TOUCH_INT_ST) {
-        for (i = 0; i < 10; ++i) {
-            if ((pad_intr >> i) & 0x01) {
-                if(__touchInterruptHandlers[i]){
-                    __touchInterruptHandlers[i]();
+    touch_pad_clear_status();
+    // call Pad ISR User callback
+    for (int i = 0; i < SOC_TOUCH_SENSOR_NUM; i++) {
+        if ((pad_intr >> i) & 0x01) {
+            if(__touchInterruptHandlers[i].fn){
+                // keeping backward compatibility with "void cb(void)" and with new "void cb(vooid *)"
+                if (__touchInterruptHandlers[i].callWithArgs) {
+                    ((voidArgFuncPtr)__touchInterruptHandlers[i].fn)(__touchInterruptHandlers[i].arg);
+                } else {
+                    __touchInterruptHandlers[i].fn();
                 }
             }
         }
     }
 #endif
+
+#if SOC_TOUCH_VERSION_2     // ESP32S2, ESP32S3
+            touch_pad_intr_mask_t evt = touch_pad_read_intr_status_mask();
+            uint8_t pad_num = touch_pad_get_current_meas_channel();
+            if (evt & TOUCH_PAD_INTR_MASK_ACTIVE) {
+                // touch has been pressed / touched
+                __touchInterruptHandlers[pad_num].lastStatusIsPressed = true;
+            }
+            if (evt & TOUCH_PAD_INTR_MASK_INACTIVE) {
+                // touch has been released / untouched
+                __touchInterruptHandlers[pad_num].lastStatusIsPressed = false;
+            }
+            if(__touchInterruptHandlers[pad_num].fn){
+                // keeping backward compatibility with "void cb(void)" and with new "void cb(vooid *)"
+                if (__touchInterruptHandlers[pad_num].callWithArgs) {
+                    ((voidArgFuncPtr)__touchInterruptHandlers[pad_num].fn)(__touchInterruptHandlers[pad_num].arg);
+                } else {
+                    __touchInterruptHandlers[pad_num].fn();
+                }
+            }
+#endif
 }
 
-void __touchSetCycles(uint16_t measure, uint16_t sleep)
+
+
+static void __touchSetCycles(uint16_t measure, uint16_t sleep)
 {
     __touchSleepCycles = sleep;
     __touchMeasureCycles = measure;
-#if CONFIG_IDF_TARGET_ESP32
-    //Touch pad SleepCycle Time
-    SET_PERI_REG_BITS(SENS_SAR_TOUCH_CTRL2_REG, SENS_TOUCH_SLEEP_CYCLES, __touchSleepCycles, SENS_TOUCH_SLEEP_CYCLES_S);
-    //Touch Pad Measure Time
-    SET_PERI_REG_BITS(SENS_SAR_TOUCH_CTRL1_REG, SENS_TOUCH_MEAS_DELAY, __touchMeasureCycles, SENS_TOUCH_MEAS_DELAY_S);
-#else
+#if SOC_TOUCH_VERSION_1  || SOC_TOUCH_VERSION_2         // ESP32 || ESP32S2, ESP32S3
```

Comment:
This can be removed. It has no effect.

Reply:
I Agree that both #IF cover all cases and thus they have no effect, but I decided to do it to make the code more explicit for the future, in case IDF comes up with a Version 3 for other chip, for instance.

What do you say?

Reply:
Yes, that makes a sense.  Its can be actually good when new chip comes with new Version 3, so the touch won't be working at all. There will be easily documented places where to look at changes 👍